firmware/applications/SignBuddy/uart.c

```diff
@@ -0,0 +1,158 @@
+#include "uart.h"
+
+#include "board.h"
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <string.h>
+
+#define IS_POWER_OF_TWO(num) (((num) & ((num) - 1)) == 0) ? true : false
+
+#define ASSERT(val)         \
+  do {                      \
+    if(!(val)) {            \
+      error_handler();      \
+    }                       \
+  } while (0)
+
+typedef struct {
+  uint8_t *buffer;
+  uint32_t write_index;
+  uint32_t read_index;
+  uint32_t bytes_available;
+  uint32_t size;
+} fifo_t;
+
+static uint8_t rx_buffer[UART_RX_BUFFER_SIZE];
+static uint8_t tx_buffer[UART_TX_BUFFER_SIZE];
+
+static fifo_t rx_fifo;
+static fifo_t tx_fifo;
+
+static bool writing = false;
+
+static void fifo_init(fifo_t *fifo, uint8_t *buffer, uint32_t size)
+{
+  ASSERT(IS_POWER_OF_TWO(size));
+  fifo->buffer = buffer;
+  fifo->size = size;
+  fifo->write_index = 0;
+  fifo->read_index = 0;
+  fifo->bytes_available = 0;
+}
+
+static void fifo_push(fifo_t *fifo, uint8_t *buffer, uint32_t length)
+{
+  for (uint32_t i = 0; i < length; i++) {
+    fifo->buffer[fifo->write_index] = buffer[i];
+    fifo->write_index = (fifo->write_index + 1) & (fifo->size - 1);
+  }
+
+  fifo->bytes_available = (fifo->bytes_available + length) & (fifo->size - 1);
+}
+
+static int fifo_pop(fifo_t *fifo, uint8_t *dest, uint32_t length)
+{
+  if (length > fifo->bytes_available) {
+    return 0;
+  }
+
+  for (uint32_t i = 0; i < length; i++) {
+    dest[i] = fifo->buffer[fifo->read_index];
+    fifo->read_index = (fifo->read_index + 1) & (fifo->size - 1);
+  }
+  fifo->bytes_available -= length;
+  return length;
+}
+
+static void hw_init(void)
+{
+  LL_USART_InitTypeDef uart_config = {0};
+  LL_GPIO_InitTypeDef gpio_config = {0};
+
+  UART_CLK_EN();
+  GPIO_UART_CLK_EN();
+
+  gpio_config.Pin = UART_RX_PIN | UART_TX_PIN;
+  gpio_config.Mode = LL_GPIO_MODE_ALTERNATE;
+  gpio_config.Speed = LL_GPIO_SPEED_FREQ_VERY_HIGH;
+  gpio_config.OutputType = LL_GPIO_OUTPUT_PUSHPULL;
+  gpio_config.Pull = LL_GPIO_PULL_NO;
+  gpio_config.Alternate = UART_GPIO_AF;
+  LL_GPIO_Init(UART_GPIO_PORT, &gpio_config);
+
+  uart_config.BaudRate = UART_BAUDRATE;
+  uart_config.DataWidth = LL_USART_DATAWIDTH_8B;
+  uart_config.StopBits = LL_USART_STOPBITS_1;
+  uart_config.Parity = LL_USART_PARITY_NONE;
+  uart_config.TransferDirection = LL_USART_DIRECTION_TX_RX;
+  uart_config.HardwareFlowControl = LL_USART_HWCONTROL_NONE;
+  uart_config.OverSampling = LL_USART_OVERSAMPLING_16;
+
+  LL_USART_Init(UART, &uart_config);
+
+  LL_USART_ConfigAsyncMode(UART);
+  LL_USART_Enable(UART);
+
+  LL_USART_EnableIT_RXNE(UART);
+  LL_USART_EnableIT_TC(UART);
+
+  NVIC_SetPriority(UART_IRQn, 0);
```

Comment:
Define priority with macro

Reply:
what is the reasoning behind using that macro?

Reply:
I forget how the nvic priorities are encoded, but I seem to remember it wasn't too hard to keep track of in my head.

anyway, I think we should use it to keep things consistent

Reply:
Reasoning would be just to make it easier to keep track of if we want to change all the priorities